subcommands/targets/add.go

```diff
@@ -0,0 +1,288 @@
+package targets
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"github.com/theupdateframework/notary"
+	tuf "github.com/theupdateframework/notary/tuf/data"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	addTargetType string
+	addTags       string
+	addSrcTag     string
+	addOutFile    string
+	addPrint      bool
+	addDryRun     bool
+)
+
+// Consider moving the `Target` type along with its receivers to the `client` package
+// or do opposite - "hide" it
+type (
+	Target struct {
+		Length int64             `json:"length"`
+		Hashes tuf.Hashes        `json:"hashes"`
+		Custom *client.TufCustom `json:"custom,omitempty"`
+	}
+	Targets map[string]*Target
+)
+
+func NewTarget(tufMeta tuf.FileMeta) (*Target, error) {
+	custom, err := api.TargetCustom(tufMeta)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal Target custom from TUF meta: %s", err.Error())
+	}
+	return &Target{
+		Length: tufMeta.Length,
+		Hashes: tufMeta.Hashes,
+		Custom: custom,
+	}, nil
+}
+
+func (t *Target) Derive(newVer int) *Target {
+	derivedCustom := *t.Custom
+	derivedCustom.Version = strconv.Itoa(newVer)
+	derivedCustom.CreatedAt = time.Now().UTC().Format(time.RFC3339)
+	derivedCustom.UpdatedAt = derivedCustom.CreatedAt
+	return &Target{
+		Length: t.Length,
+		Hashes: t.Hashes,
+		Custom: &derivedCustom,
+	}
+}
+
+func (t *Target) HardwareId() string {
+	return t.Custom.HardwareIds[0]
+}
+
+func (t *Target) Version() int {
+	ver, err := strconv.Atoi(t.Custom.Version)
+	if err != nil {
+		panic(fmt.Sprintf("failed to convert Target version to int: %s", err.Error()))
+	}
+	return ver
+}
+
+func (t *Target) Tags() []string {
+	return t.Custom.Tags
+}
+
+func (t *Target) HasTag(tag string) bool {
+	for _, ct := range t.Tags() {
+		if ct == tag {
+			return true
+		}
+	}
+	return false
+}
+
+func (t *Target) Name() string {
+	return t.Custom.Name + "-" + t.Custom.Version
+}
+
+func (t *Target) SetHash(hash string) error {
+	hb, err := base64.StdEncoding.DecodeString(hash)
+	if err != nil {
+		return err
+	}
+	t.Hashes[notary.SHA256] = hb
+	return nil
+}
+
+func init() {
+	addCmd := &cobra.Command{
+		Use:   "add",
+		Short: "Add target(s)",
+		Run:   doAdd,
+		Long:  "",
+	}
+	cmd.AddCommand(addCmd)
+	addCmd.Flags().StringVarP(&addTargetType, "type", "", "", "Target type")
+	addCmd.Flags().StringVarP(&addTags, "tags", "", "", "comma,separate,list of Target tags")
+	addCmd.Flags().StringVarP(&addSrcTag, "src-tag", "", "", "OSTree Target tag to base app targets on")
+	addCmd.Flags().StringVarP(&addOutFile, "out", "o", "", "a file path to output the new Targets to")
```

Comment:
At first this felt like extra functionality that nobody is asking for. However, are you adding this so that it could be used by our CI and we'd set this value to `/archive/....` to get debug info from yocto someday?

Reply:
Yes, as well as it may be used by customer CI too.

Reply:
I removed it, will add it back on a request.